pkg/controller/miganalytic/miganalytics_controller.go

```diff
@@ -0,0 +1,582 @@
+	analytic.MarkReconciled()
+	err = r.Update(context.TODO(), analytic)
+	if err != nil {
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Done
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileMigAnalytic) analyze(analytic *migapi.MigAnalytic) error {
+	plan := &migapi.MigPlan{}
+
+	err := r.getAnalyticPlan(analytic, plan)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	srcCluster, err := plan.GetSourceCluster(r)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	client, err := srcCluster.GetClient(r)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	discovery, err := discovery.NewDiscoveryClientForConfig(client.RestConfig())
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	resources, err := collectResources(discovery)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	dynamic, err := dynamic.NewForConfig(client.RestConfig())
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	analytic.Status.Analytics.Plan = plan.Name
+	log.Info("Beginning analysis of plan.", "migplan", plan.Name)
+
+	for i, namespace := range plan.Spec.Namespaces {
+		log.Info("Beginning analysis of namespace.", "namespace", namespace)
+		ns := migapi.MigAnalyticNamespace{
+			Namespace: namespace,
+		}
+
+		if analytic.Spec.AnalyzeK8SResources {
+			err := r.analyzeK8SResources(dynamic, resources, &ns, plan.Status.ExcludedResources, plan.Status.Incompatible)
+			if err != nil {
+				log.Trace(err)
+				return err
+			}
+		}
+		if analytic.Spec.AnalyzeImageCount {
+			err := r.analyzeImages(client, &ns, analytic.Spec.ListImages, analytic.Spec.MaxListImages)
+			if err != nil {
+				log.Trace(err)
+				return err
+			}
+		}
+		if analytic.Spec.AnalyzePVCapacity {
+			err := r.analyzePVCapacity(client, &ns)
+			if err != nil {
+				log.Trace(err)
+				return err
+			}
+		}
+
+		analytic.Status.Analytics.Namespaces = append(analytic.Status.Analytics.Namespaces, ns)
+		analytic.Status.Analytics.K8SResourceTotal += ns.K8SResourceTotal
+		analytic.Status.Analytics.ExcludedK8SResourceTotal += ns.ExcludedK8SResourceTotal
+		analytic.Status.Analytics.IncompatibleK8SResourceTotal += ns.IncompatibleK8SResourceTotal
+		analytic.Status.Analytics.ImageCount += ns.ImageCount
+		analytic.Status.Analytics.ImageSizeTotal.Add(ns.ImageSizeTotal)
+		analytic.Status.Analytics.PVCapacity.Add(ns.PVCapacity)
+		analytic.Status.Analytics.PVCount += ns.PVCount
+		analytic.Status.Analytics.PercentComplete = (i + 1) * 100 / len(plan.Spec.Namespaces)
+
+		err = r.Update(context.TODO(), analytic)
+		if err != nil {
+			log.Trace(err)
+			return err
+		}
+		log.Info("Completed analysis of namespace.", "namespace", namespace)
+	}
+	log.Info("Completed analysis of plan.", "plan", plan.Name)
+	return nil
+}
+
+func (r *ReconcileMigAnalytic) analyzeK8SResources(dynamic dynamic.Interface, resources []*metav1.APIResourceList, namespace *migapi.MigAnalyticNamespace, excludedResources []string, incompatible v1alpha1.Incompatible) error {
```

Comment:
Can we get a comment summary of what all is happening in this method? Let's consider decomposing this func to bring down line lengths.

Reply:
ya, it grew a bit more than I expected when I added a the list of incompatible and excluded resources.

It's basically comparing GVKs against the two lists that will cause them to be exempted from the migration and placing them in the right list:
```
  excludedResources:
  - imagetags
  - templateinstances
  - clusterserviceversions
  - packagemanifests
  - subscriptions
  - servicebrokers
  - servicebindings
  - serviceclasses
  - serviceinstances
  - serviceplans
  incompatibleNamespaces:
  - gvks:
    - group: apps
      kind: deployments
      version: v1beta1
    - group: extensions
      kind: replicasets
      version: v1beta1
```

resulting in something like:
```
    - excludedK8SResourceTotal: 355
      excludedK8SResources:
      - count: 355
        group: packages.operators.coreos.com
        kind: packagemanifests
        version: v1
...
      k8sResources:
      - count: 2
        kind: endpoints
        version: v1
      - count: 2
        kind: persistentvolumeclaims
        version: v1
...
```

The incompatibleResourcess is empty/omitted in the example above. I don't have a cluster stood up at the moment with any incompatible GVKs, but they would appare similarly

Reply:
pushed some of it out to separate functions, added comments, hopefully it's more readable now.